File: go/blackroad.go
// Package blackroad provides a Go client for the BlackRoad OS API.
//
// Usage:
//
//	client := blackroad.New()
//	status, err := client.Fleet.Status()
//	err = client.Slack.Post("hello from go")
package blackroad

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type Client struct {
	StatsURL   string
	SlackURL   string
	GatewayURL string
	HTTP       *http.Client
	Fleet      *FleetService
	Slack      *SlackService
	AI         *AIService
}

func New(opts ...Option) *Client {
	c := &Client{
		StatsURL:   "https://stats-blackroad.amundsonalexa.workers.dev",
		SlackURL:   "https://blackroad-slack.amundsonalexa.workers.dev",
		GatewayURL: "http://localhost:11434",
		HTTP:       &http.Client{Timeout: 30 * time.Second},
	}
	for _, opt := range opts {
		opt(c)
	}
	c.Fleet = &FleetService{client: c}
	c.Slack = &SlackService{client: c}
	c.AI = &AIService{client: c}
	return c
}

type Option func(*Client)

func WithStatsURL(url string) Option   { return func(c *Client) { c.StatsURL = strings.TrimRight(url, "/") } }
func WithSlackURL(url string) Option   { return func(c *Client) { c.SlackURL = strings.TrimRight(url, "/") } }
func WithGatewayURL(url string) Option { return func(c *Client) { c.GatewayURL = strings.TrimRight(url, "/") } }

// Fleet

type FleetService struct{ client *Client }

func (f *FleetService) Status() (map[string]interface{}, error) {
	return f.client.get(f.client.StatsURL + "/fleet")
}

func (f *FleetService) All() (map[string]interface{}, error) {
	return f.client.get(f.client.StatsURL + "/all")
}

func (f *FleetService) Health() (map[string]interface{}, error) {
	return f.client.get(f.client.StatsURL + "/health")
}

// Slack

type SlackService struct{ client *Client }

func (s *SlackService) Post(text string) (map[string]interface{}, error) {
	return s.client.post(s.client.SlackURL+"/post", map[string]string{"text": text})
}

func (s *SlackService) Alert(text string) (map[string]interface{}, error) {
	return s.client.post(s.client.SlackURL+"/alert", map[string]string{"text": text})
}

func (s *SlackService) Status() (map[string]interface{}, error) {
	return s.client.get(s.client.SlackURL + "/status")
}

// AI

type AIService struct{ client *Client }

func (a *AIService) Generate(model, prompt string) (map[string]interface{}, error) {
	return a.client.post(a.client.GatewayURL+"/api/generate", map[string]interface{}{
		"model": model, "prompt": prompt, "stream": false,
	})
}

func (a *AIService) Models() (map[string]interface{}, error) {
	return a.client.get(a.client.GatewayURL + "/api/tags")
}

// HTTP helpers

func (c *Client) get(url string) (map[string]interface{}, error) {
	resp, err := c.HTTP.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	return decodeJSON(resp.Body)
}

func (c *Client) post(url string, data interface{}) (map[string]interface{}, error) {
	body, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}
	resp, err := c.HTTP.Post(url, "application/json", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	return decodeJSON(resp.Body)
}

func decodeJSON(r io.Reader) (map[string]interface{}, error) {
	var result map[string]interface{}
	if err := json.NewDecoder(r).Decode(&result); err != nil {
		return nil, fmt.Errorf("decode: %w", err)
	}
	return result, nil
}

File: python/blackroad.py
"""
BlackRoad OS Python SDK

Connect to BlackRoad fleet, AI inference, stats, and Slack hub.

    from blackroad import BlackRoad

    br = BlackRoad()
    status = br.fleet.status()
    br.slack.post("hello from python")
"""

import urllib.request
import urllib.error
import json
from typing import Optional


class _Fleet:
    """Fleet status and health monitoring."""

    def __init__(self, client):
        self._client = client

    def status(self) -> dict:
        """Get fleet status for all nodes."""
        return self._client._get(f"{self._client.stats_url}/fleet")

    def all(self) -> dict:
        """Get all fleet data: infra, GitHub, analytics."""
        return self._client._get(f"{self._client.stats_url}/all")

    def health(self) -> dict:
        """Health check for the stats API."""
        return self._client._get(f"{self._client.stats_url}/health")


class _Slack:
    """Post messages to Slack through the BlackRoad hub."""

    def __init__(self, client):
        self._client = client

    def post(self, text: str) -> dict:
        """Post a message to the default Slack channel."""
        return self._client._post(f"{self._client.slack_url}/post", {"text": text})

    def alert(self, text: str) -> dict:
        """Post an alert to the alerts channel."""
        return self._client._post(f"{self._client.slack_url}/alert", {"text": text})

    def deploy(self, text: str) -> dict:
        """Post a deploy notification."""
        return self._client._post(f"{self._client.slack_url}/deploy", {"text": text})

    def status(self) -> dict:
        """Get Slack hub status."""
        return self._client._get(f"{self._client.slack_url}/status")


class _AI:
    """AI inference through the BlackRoad gateway."""

    def __init__(self, client):
        self._client = client

    def generate(self, model: str, prompt: str, stream: bool = False) -> dict:
        """Generate a completion using Ollama on the fleet."""
        return self._client._post(
            f"{self._client.gateway_url}/api/generate",
            {"model": model, "prompt": prompt, "stream": stream},
        )

    def chat(self, model: str, messages: list, stream: bool = False) -> dict:
        """Chat completion using Ollama on the fleet."""
        return self._client._post(
            f"{self._client.gateway_url}/api/chat",
            {"model": model, "messages": messages, "stream": stream},
        )

    def embed(self, model: str, input: str) -> dict:
        """Generate embeddings."""
        return self._client._post(
            f"{self._client.gateway_url}/api/embed",
            {"model": model, "input": input},
        )

    def models(self) -> dict:
        """List available models."""
        return self._client._get(f"{self._client.gateway_url}/api/tags")


class BlackRoad:
    """BlackRoad OS SDK client.

    Args:
        stats_url: Stats API URL (default: public endpoint)
        slack_url: Slack hub URL (default: public endpoint)
        gateway_url: AI gateway URL (default: localhost Ollama)
    """

    def __init__(
        self,
        stats_url: str = "https://stats-blackroad.amundsonalexa.workers.dev",
        slack_url: str = "https://blackroad-slack.amundsonalexa.workers.dev",
        gateway_url: str = "http://localhost:11434",
    ):
        self.stats_url = stats_url.rstrip("/")
        self.slack_url = slack_url.rstrip("/")
        self.gateway_url = gateway_url.rstrip("/")

        self.fleet = _Fleet(self)
        self.slack = _Slack(self)
        self.ai = _AI(self)

    def _get(self, url: str) -> dict:
        req = urllib.request.Request(url)
        try:
            with urllib.request.urlopen(req, timeout=10) as resp:
                return json.loads(resp.read().decode())
        except urllib.error.URLError as e:
            return {"error": str(e)}

    def _post(self, url: str, data: dict) -> dict:
        body = json.dumps(data).encode()
        req = urllib.request.Request(
            url, data=body, headers={"Content-Type": "application/json"}, method="POST"
        )
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                return json.loads(resp.read().decode())
        except urllib.error.URLError as e:
            return {"error": str(e)}


if __name__ == "__main__":
    br = BlackRoad()
    print("Fleet:", json.dumps(br.fleet.health(), indent=2))
    print("Slack:", json.dumps(br.slack.status(), indent=2))
